src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -820,34 +847,53 @@ private static LevelsClause ParseLevels(long? levelsOption, IEdmType sourceType,
+            IEdmNavigationSource resourcePathNavigationSource,
             IEdmNavigationSource targetNavigationSource,
             IEdmTypeReference elementType,
             HashSet<EndPathToken> generatedProperties = null,
             bool collapsed = false)
         {
-            BindingState state = CreateBindingState(config, targetNavigationSource, elementType, generatedProperties, collapsed);
+            BindingState state = CreateBindingState(config, resourcePathNavigationSource, targetNavigationSource, elementType, generatedProperties, collapsed);
             return new MetadataBinder(state);
         }
 
         private static BindingState CreateBindingState(ODataUriParserConfiguration config,
-            IEdmNavigationSource targetNavigationSource, IEdmTypeReference elementType,
-            HashSet<EndPathToken> generatedProperties = null, bool collapsed = false)
+            IEdmNavigationSource resourcePathNavigationSource, IEdmNavigationSource targetNavigationSource,
+            IEdmTypeReference elementType, HashSet<EndPathToken> generatedProperties = null,
+            bool collapsed = false)
         {
             if (targetNavigationSource == null && elementType == null)
             {
                 return null;
             }
 
+            // For example if we have https://url/Books?$expand=Authors($filter=Name eq $it/Name)
+            // $filter=Name will reference Authors(the expanded entity).
+            // $it/Name will reference Books(the resource identified by the path).
+            // The BindingState ImplicitRangeVariable property will store the $it that references the expanded/selected item (The Implicit Range Variable).
+            // We the add to the Stack, the $it that references the resource identified by the path (The Explicit Range Variable).
+
             BindingState state = new BindingState(config)
             {
                 ImplicitRangeVariable =
                     NodeFactory.CreateImplicitRangeVariable(elementType != null ? elementType :
                         targetNavigationSource.EntityType().ToTypeReference(), targetNavigationSource)
             };
 
-            state.RangeVariables.Push(state.ImplicitRangeVariable);
             state.AggregatedPropertyNames = generatedProperties;
             state.IsCollapsed = collapsed;
+            state.ResourcePathNavigationSource = resourcePathNavigationSource;
+
+            if (resourcePathNavigationSource != null)
+            {
+                // This $it rangeVariable will be added the Stack.
+                // We are adding a rangeVariable whose navigationSource is the resource path entity set.
+                // ODATA spec: Example 106 http://docs.oasis-open.org/odata/odata/v4.01/csprd05/part2-url-conventions/odata-v4.01-csprd05-part2-url-conventions.html#sec_it
+                RangeVariable explicitRangeVariable = NodeFactory.CreateImplicitRangeVariable(

```

Comment:
We call `NodeFactory.CreateImplicitRangeVariable()` here and assign it to a variable named `explicitRangeVariable`... Is use of "explicit" in the variable name deliberate?

Reply:
It's deliberate --
We use the same Factory method to create both the implicit and explicit range variables.